### PR TITLE
Rework PLT exclusion code

### PR DESCRIPTION
**Pull request by WorksButNotTested**

### CODE REVIEW

File: gum/backend-x86/gumstalker-x86.c [hunk 1]
```diff
@@ -4625,91 +4646,114 @@ gum_exec_block_is_direct_jmp_to_plt_got (GumExecBlock * block,
...
-    goto beach;
+    return FALSE;
 
-  map = gum_module_map_new ();
+  ranges = gum_exec_ctx_get_plt_got_ranges ();
 
-  module = gum_module_map_find (map, GUM_ADDRESS (target->absolute_address));
-  if (module == NULL)
-    goto beach;
+  for (guint i = 0; i < ranges->len; i++) {
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c:4672]:
```suggestion
  for (i = 0; i != ranges->len; i++)
  {
```

(Note that I removed the variable declaration, as variable declarations should be hoisted to the nearest scope.)

File: gum/backend-x86/gumstalker-x86.c [hunk 2]
```diff
@@ -4625,91 +4646,114 @@ gum_exec_block_is_direct_jmp_to_plt_got (GumExecBlock * block,
...
+    range = &g_array_index(ranges, GumMemoryRange, i);
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c:4673]:
```suggestion
    range = &g_array_index (ranges, GumMemoryRange, i);
```

File: gum/backend-x86/gumstalker-x86.c [hunk 3]
```diff
@@ -4625,91 +4646,114 @@ gum_exec_block_is_direct_jmp_to_plt_got (GumExecBlock * block,
...
+
+    if (GUM_MEMORY_RANGE_INCLUDES (range,
+        GPOINTER_TO_SIZE (target->absolute_address)))
+    {
+      return TRUE;
+    }
+  }
+#endif
+  return FALSE;
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c:4682]:
```suggestion

  return FALSE;
```

File: gum/backend-x86/gumstalker-x86.c [hunk 4]
```diff
@@ -4625,91 +4646,114 @@ gum_exec_block_is_direct_jmp_to_plt_got (GumExecBlock * block,
...
+gum_exec_ctx_get_plt_got_ranges (void)
+{
+  static gsize gonce_value;
 
-  elf = gum_elf_module_new_from_memory (module->path,
-      module->range->base_address);
-  g_assert (elf != NULL);
+  if (g_once_init_enter (&gonce_value))
+  {
+    GArray * ranges = g_array_new(false, false, sizeof(GumMemoryRange));
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c:4694]:
```suggestion
    GArray * ranges = g_array_new (FALSE, FALSE, sizeof (GumMemoryRange));
```

File: gum/backend-x86/gumstalker-x86.c [hunk 5]
```diff
@@ -4625,91 +4646,114 @@ gum_exec_block_is_direct_jmp_to_plt_got (GumExecBlock * block,
...
-    return FALSE;
-
-  limit = details->address + details->size;
-  if (check->target->absolute_address >= GSIZE_TO_POINTER (limit))
-    return FALSE;
-
-  check->found = TRUE;
+  range.base_address = details->address;
+  range.size = details->size;
+  g_array_append_val(ranges, range);
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c:4754]:
```suggestion
  g_array_append_val (ranges, range);
```

File: gum/backend-x86/gumstalker-x86.c [hunk 6]
```diff
@@ -4625,91 +4646,114 @@ gum_exec_block_is_direct_jmp_to_plt_got (GumExecBlock * block,
...
-    .target = target,
-    .found = FALSE
-  };
-  GumCheckElfSection plt_sec = {
-    .name = ".plt.sec",
-    .target = target,
-    .found = FALSE
-  };
+  GArray * ranges;
+  GumMemoryRange *range;
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c:4653]:
```suggestion
  GumMemoryRange * range;
```

But this should be moved to the inner scope where it is needed.

File: gum/backend-x86/gumstalker-x86.c [hunk 7]
```diff
@@ -4625,91 +4646,114 @@ gum_exec_block_is_direct_jmp_to_plt_got (GumExecBlock * block,
...
+  ranges = gum_exec_ctx_get_plt_got_ranges ();
 
-  module = gum_module_map_find (map, GUM_ADDRESS (target->absolute_address));
-  if (module == NULL)
-    goto beach;
+  for (guint i = 0; i < ranges->len; i++) {
+    range = &g_array_index(ranges, GumMemoryRange, i);
+
+    if (GUM_MEMORY_RANGE_INCLUDES (range,
+        GPOINTER_TO_SIZE (target->absolute_address)))
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c:4676]: Perhaps we should `g_array_sort()` the ranges, and do a binary search using `g_array_binary_search()`?

**Review by oleavr — CHANGES_REQUESTED** [review]
Thanks! :boom:

File: gum/backend-x86/gumstalker-x86.c (hunk 7 above)
Comment (WorksButNotTested) [reply gum/backend-x86/gumstalker-x86.c:4676]: The array should be quite small and only used on the slow path.

**Comment by WorksButNotTested** [comment]
Ok. Hopefully that should sort it.

File: gum/backend-x86/gumstalker-x86.c [hunk 8]
```diff
@@ -969,6 +973,23 @@ gum_stalker_init (GumStalker * self)
...
 
+#if defined (HAVE_LINUX)
+  /*
+   * We need to build an array of ranges in which the .plt.got and .plt.sec
+   * sections of the loaded modules reside to allow us to treat tail calls into
+   * them as excluded calls (even though they use a JMP instruction). However,
+   * calling into the dynamic loader or even just allocating data on the heap is
+   * dangerous when actually stalking a target since we could cause the target
+   * to re-enter a section of code which is not designed to be. We will
+   * therefore build up our picture of the memory map when stalker is first
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c:984]:
```suggestion
   * therefore build up our picture of the memory map when Stalker is first
```

File: gum/backend-x86/gumstalker-x86.c [hunk 9]
```diff
@@ -4625,91 +4646,118 @@ gum_exec_block_is_direct_jmp_to_plt_got (GumExecBlock * block,
...
-      module->range->base_address);
-  g_assert (elf != NULL);
+static GArray *
+gum_exec_ctx_get_plt_got_ranges (void)
+{
+  static gsize gonce_value;
+
+  if (g_once_init_enter (&gonce_value))
+  {
+    GArray * ranges = g_array_new (FALSE, FALSE, sizeof(GumMemoryRange));
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c:4698]:
```suggestion
    GArray * ranges = g_array_new (FALSE, FALSE, sizeof (GumMemoryRange));
```

File: gum/backend-x86/gumstalker-x86.c [hunk 10]
```diff
@@ -4625,91 +4646,118 @@ gum_exec_block_is_direct_jmp_to_plt_got (GumExecBlock * block,
...
+
+    gum_process_enumerate_modules (gum_exec_ctx_find_plt_got, ranges);
+
+    _gum_register_early_destructor (gum_exec_ctx_deinit_plt_got_ranges);
+
+    g_once_init_leave (&gonce_value, GPOINTER_TO_SIZE (ranges) + 1);
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c:4704]:
```suggestion
    g_once_init_leave (&gonce_value, GPOINTER_TO_SIZE (ranges));
```

File: gum/backend-x86/gumstalker-x86.c [hunk 11]
```diff
@@ -4625,91 +4646,118 @@ gum_exec_block_is_direct_jmp_to_plt_got (GumExecBlock * block,
...
+  }
+
+  return GSIZE_TO_POINTER (gonce_value - 1);
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c:4707]:
```suggestion
  return GSIZE_TO_POINTER (gonce_value);
```

File: gum/backend-x86/gumstalker-x86.c [hunk 12]
```diff
@@ -4625,91 +4646,118 @@ gum_exec_block_is_direct_jmp_to_plt_got (GumExecBlock * block,
...
+{
+  g_array_free (gum_exec_ctx_get_plt_got_ranges (), TRUE);
+}
+
+static gboolean
+gum_exec_ctx_find_plt_got (const GumModuleDetails * details,
+                           gpointer user_data)
+{
+  GArray * ranges = user_data;
+  GumElfModule * elf = NULL;
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c:4721]:
```suggestion
  GumElfModule * elf;
```

File: gum/backend-x86/gumstalker-x86.c [hunk 13]
```diff
@@ -4625,91 +4646,118 @@ gum_exec_block_is_direct_jmp_to_plt_got (GumExecBlock * block,
...
+
+  if (details->path == NULL)
+    return TRUE;
+
+  elf = gum_elf_module_new_from_memory (details->path,
+      details->range->base_address);
+
+  if (elf == NULL)
     goto beach;
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c:4730]:
```suggestion
  if (elf == NULL)
    return TRUE;
```

File: gum/backend-x86/gumstalker-x86.c [hunk 14]
```diff
@@ -4625,91 +4646,118 @@ gum_exec_block_is_direct_jmp_to_plt_got (GumExecBlock * block,
...
 
-  result = TRUE;
+  gum_elf_module_enumerate_sections (elf, gum_exec_check_elf_section, ranges);
 
 beach:
   g_clear_object (&elf);
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c:4735]:
```suggestion
  g_object_unref (elf);
```

**Review by oleavr — CHANGES_REQUESTED** [review]

**Comment by WorksButNotTested** [comment]
Ready for you again. Thanks.

**Pull request merged**